store/src/test/java/com/dropbox/android/external/store4/impl/ClearAllStoreTest.kt

```diff
@@ -0,0 +1,93 @@
+package com.dropbox.android.external.store4.impl
+
+import com.dropbox.android.external.store4.ExperimentalStoreApi
+import com.dropbox.android.external.store4.StoreBuilder
+import com.dropbox.android.external.store4.get
+import com.google.common.truth.Truth.assertThat
+import com.nhaarman.mockitokotlin2.mock
+import com.nhaarman.mockitokotlin2.whenever
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import java.util.concurrent.atomic.AtomicInteger
+
+@FlowPreview
+@ExperimentalCoroutinesApi
+@ExperimentalStoreApi
+@RunWith(JUnit4::class)
+class ClearAllStoreTest {
+
+    private val testScope = TestCoroutineScope()
+
+    private val key = "key"
+
+    private val networkCalls = AtomicInteger(0)
+    private val deleteAllCalls = AtomicInteger(0)
+
+    private val diskReader = mock<(String) -> Flow<Int?>>()
+    private val diskWriter = mock<suspend (String, Int) -> Unit>()
+
+    @Test
+    fun `clearAll() with persister`() = testScope.runBlockingTest {
+        whenever(diskReader.invoke(key))
+            .thenReturn(flowOf(null)) // no value in disk initially
+            .thenReturn(flowOf(1)) // value exists after fetching from network
+            .thenReturn(flowOf(null)) // no value after clearing
+            .thenReturn(flowOf(1)) // value exists after additional network call
+        whenever(diskWriter.invoke(key, 1)).thenReturn(Unit)
```

Comment:
1) In general try to avoid using mocks when you can just as easily create test classes.
2) In this case, the use of mocks is dangerous. You are actually mocking the effect that you are testing for then testing store behaves correctly based on this specific values from the persister.

This test actually has no dependency on whether you actually clear data from the presister. If you remove the clear line the test will still pass.

instead of mocking, create an in-memory source of truth (e.g a hash map+ channels to notify on changes to keys). then run your test and validate on the content of this source of truth.

Reply:
edit - you do validate specifically that `deleteAll` was called, but that's still fragile as all the other elements of the test are not dependent on it